codes/individual_definitions/individual_evaluate.py

```diff
@@ -406,10 +407,18 @@ def evaluate(self,
                 training_datalist, validating_datalist, supplements = block_material.output
         
         # Train the refiner and discriminator
-        refiner, discriminator, train_config = supplements
-        self.train_graph(indiv_material, training_datalist[0], validating_datalist[0], refiner, discriminator, train_config)
+        untrained_refiner, untrained_discriminator, train_config = supplements
+        refiners, discriminators = self.train_graph(indiv_material, training_datalist[0], validating_datalist[0], 
+            untrained_refiner, untrained_discriminator, train_config)
+
+        # Now do tournament selection to pick the best refiner/discriminator networks from the training process
+        # TODO: consider finding a way to replace this with a convergence metric or something
+        ezLogging.debug("Finding best refiner and discriminator for individual")
+        refiner_ratings, discriminator_ratings = get_graph_ratings(refiners, discriminators, validating_datalist[0], train_config['device'])
+        best_refiner = refiners[refiner_ratings['r'].argmax()]
+        best_discriminator = discriminators[discriminator_ratings['r'].argmax()]
         
-        indiv_material.output = last_output
+        indiv_material.output = (best_refiner, best_discriminator)
```

Comment:
Currently, for each individual in a population, we save the refiner and discriminator every N steps (currently set to 10, but that's just for testing, will probably be 100), and then at the end, we pass all refiners and discriminators into the tournament evaluation method (called with `get_graph_ratings`). This gives us the ratings for all the refiners and discriminators and then we just take the refiner and discriminator that performed best as our output for the individual.